File: src/java/libs/core/seqvar/SeqVar.java
package libs.core.seqvar;

import java.io.File;
import java.sql.*;
import java.time.Instant;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * Java implementation of the SeqVar store - a simple key-value configuration system
 * for SeqWeb development environment.
 * 
 * This class provides persistent storage of configuration values with source tracking
 * and timestamps, similar to the Python seqvar implementation.
 */
public class SeqVar {
    
    /**
     * Exception thrown by SeqVar operations
     */
    public static class SeqVarException extends RuntimeException {
        public SeqVarException(String message) {
            super(message);
        }
        
        public SeqVarException(String message, Throwable cause) {
            super(message, cause);
        }
    }
    
    private static final String DB_NAME = "seqvar.sqlite";
    private static final String SCHEMA_SQL = """
        PRAGMA journal_mode=WAL;
        PRAGMA synchronous=NORMAL;
        CREATE TABLE IF NOT EXISTS seqvars(
          key  TEXT NOT NULL,
          val  TEXT NOT NULL DEFAULT '',
          src  TEXT,
          ts   INTEGER NOT NULL,
          PRIMARY KEY(key)
        );
        """;
    
    /**
     * Get a seqvar value by key
     * 
     * @param key the configuration key
     * @return the value, or null if not found
     * @throws SeqVarException if there's a database error
     */
    public static String get(String key) throws SeqVarException {
        try (Connection conn = conn();
             PreparedStatement stmt = conn.prepareStatement(
                 "SELECT val FROM seqvars WHERE key = ?")) {
            
            stmt.setString(1, key);
            ResultSet rs = stmt.executeQuery();
            
            if (rs.next()) {
                return rs.getString("val");
            }
            return null;
            
        } catch (SQLException e) {
            throw new SeqVarException("Error getting seqvar '" + key + "': " + e.getMessage(), e);
        }
    }
    
    /**
     * Set a seqvar value
     * 
     * @param key the configuration key
     * @param value the value to set
     * @param src the source of this value (can be null to auto-detect)
     * @throws SeqVarException if there's a database error
     */
    public static void set(String key, String value, String src) throws SeqVarException {
        if (key == null || key.trim().isEmpty()) {
            throw new SeqVarException("Key cannot be null or empty");
        }
        
        if (value == null) {
            value = "";
        }
        
        // Auto-detect source if not provided
        if (src == null) {
            src = inferSource();
        }
        
        long timestamp = Instant.now().getEpochSecond();
        
        try (Connection conn = conn();
             PreparedStatement stmt = conn.prepareStatement(
                 "INSERT OR REPLACE INTO seqvars (key, val, src, ts) VALUES (?, ?, ?, ?)")) {
            
            stmt.setString(1, key);
            stmt.setString(2, value);
            stmt.setString(3, src);
            stmt.setLong(4, timestamp);
            
            stmt.executeUpdate();
            
        } catch (SQLException e) {
            throw new SeqVarException("Error setting seqvar '" + key + "': " + e.getMessage(), e);
        }
    }
    
    /**
     * Set a seqvar value with auto-detected source
     * 
     * @param key the configuration key
     * @param value the value to set
     * @throws SeqVarException if there's a database error
     */
    public static void set(String key, String value) throws SeqVarException {
        set(key, value, null);
    }
    
    /**
     * Get all seqvar entries as a list of arrays [key, value, source, timestamp]
     * 
     * @return list of entries
     * @throws SeqVarException if there's a database error
     */
    public static List<Object[]> dump() throws SeqVarException {
        List<Object[]> results = new ArrayList<>();
        
        try (Connection conn = conn();
             PreparedStatement stmt = conn.prepareStatement(
                 "SELECT key, val, src, ts FROM seqvars ORDER BY key")) {
            
            ResultSet rs = stmt.executeQuery();
            
            while (rs.next()) {
                Object[] row = {
                    rs.getString("key"),
                    rs.getString("val"),
                    rs.getString("src"),
                    rs.getLong("ts")
                };
                results.add(row);
            }
            
        } catch (SQLException e) {
            throw new SeqVarException("Error dumping seqvars: " + e.getMessage(), e);
        }
        
        return results;
    }
    
    /**
     * Get seqvar entries as a Map, optionally filtered by pattern
     * 
     * @param pattern SQL LIKE pattern (e.g., "repos.*"), or null for all
     * @return map of key-value pairs
     * @throws SeqVarException if there's a database error
     */
    public static Map<String, String> getDict(String pattern) throws SeqVarException {
        Map<String, String> results = new HashMap<>();
        
        String sql = "SELECT key, val FROM seqvars";
        if (pattern != null && !pattern.trim().isEmpty()) {
            sql += " WHERE key LIKE ?";
        }
        sql += " ORDER BY key";
        
        try (Connection conn = conn();
             PreparedStatement stmt = conn.prepareStatement(sql)) {
            
            if (pattern != null && !pattern.trim().isEmpty()) {
                stmt.setString(1, pattern);
            }
            
            ResultSet rs = stmt.executeQuery();
            
            while (rs.next()) {
                results.put(rs.getString("key"), rs.getString("val"));
            }
            
        } catch (SQLException e) {
            throw new SeqVarException("Error getting seqvar dict: " + e.getMessage(), e);
        }
        
        return results;
    }
    
    /**
     * Get all seqvar entries as a Map
     * 
     * @return map of all key-value pairs
     * @throws SeqVarException if there's a database error
     */
    public static Map<String, String> getDict() throws SeqVarException {
        return getDict(null);
    }
    
    /**
     * Get the path to the seqvar store database
     * 
     * @return the database file path
     */
    public static File seqvarStorePath() {
        String home = System.getenv("SEQWEBDEV_HOME");
        if (home == null) {
            throw new SeqVarException("SEQWEBDEV_HOME environment variable is not set");
        }
        
        File stateDir = new File(home, ".state");
        return new File(stateDir, DB_NAME);
    }
    
    /**
     * Get a database connection, creating the database and schema if needed
     * 
     * @return database connection
     * @throws SQLException if connection fails
     */
    private static Connection conn() throws SQLException {
        File dbFile = seqvarStorePath();
        File dbDir = dbFile.getParentFile();
        
        // Create directory if it doesn't exist
        if (!dbDir.exists() && !dbDir.mkdirs()) {
            throw new SeqVarException("Could not create directory: " + dbDir);
        }
        
        // Create database and schema if it doesn't exist
        if (!dbFile.exists()) {
            createDatabase(dbFile);
        }
        
        return DriverManager.getConnection("jdbc:sqlite:" + dbFile.getAbsolutePath());
    }
    
    /**
     * Create the database and schema
     * 
     * @param dbFile the database file to create
     * @throws SeqVarException if creation fails
     */
    private static void createDatabase(File dbFile) throws SeqVarException {
        try (Connection conn = DriverManager.getConnection("jdbc:sqlite:" + dbFile.getAbsolutePath())) {
            
            // Apply schema
            for (String stmt : SCHEMA_SQL.split(";")) {
                String trimmed = stmt.trim();
                if (!trimmed.isEmpty()) {
                    try (Statement sqlStmt = conn.createStatement()) {
                        sqlStmt.execute(trimmed);
                    }
                }
            }
            
        } catch (SQLException e) {
            throw new SeqVarException("Failed to create database: " + e.getMessage(), e);
        }
    }
    
    /**
     * Infer the source from the current stack trace
     * 
     * @return the inferred source name
     */
    private static String inferSource() {
        StackTraceElement[] stack = Thread.currentThread().getStackTrace();
        
        // Skip the first few frames (getStackTrace, inferSource, set)
        for (int i = 3; i < stack.length; i++) {
            StackTraceElement frame = stack[i];
            String className = frame.getClassName();
            
            // Skip internal Java classes and our own SeqVar class
            if (!className.startsWith("java.") && 
                !className.startsWith("lib.seqvar.SeqVar") &&
                !className.startsWith("sun.")) {
                
                // Extract just the class name without package
                String[] parts = className.split("\\.");
                return parts[parts.length - 1];
            }
        }
        
        return "unknown";
    }
}

File: src/python/libs/core/seqvar/seqvar.py
# seqvar store untilities

from __future__ import annotations
import sqlite3
import time
import inspect
from pathlib import Path
from libs.core.util import get_call_trace
# Removed import to break circular dependency


def seqvar_store_path() -> Path:
    """Get the path to the seqvar store database file."""
    import os
    from pathlib import Path
    
    # Get SEQWEBDEV_HOME and construct .state/seqvar/seqvar.db path
    seqwebdev_home = os.environ.get('SEQWEBDEV_HOME')
    if not seqwebdev_home:
        raise RuntimeError("SEQWEBDEV_HOME environment variable is not set!")
    
    home_path = Path(seqwebdev_home).expanduser().resolve()
    return home_path / ".state" / "seqvar" / "seqvar.db"


# Database schema for seqvar store
SCHEMA_SQL = """
PRAGMA journal_mode=WAL;
PRAGMA synchronous=NORMAL;
CREATE TABLE IF NOT EXISTS seqvars(
  key  TEXT NOT NULL,
  val  TEXT NOT NULL DEFAULT '',
  src  TEXT,
  ts   INTEGER NOT NULL,
  PRIMARY KEY(key)
);
"""


def init_seqvar_store() -> Path:
    """
    Idempotently create the seqvar store and schema at:
      $SEQWEBDEV_HOME/.state/seqvar.sqlite

    - Creates the .state directory if needed
    - Enables WAL journaling
    - Ensures the seqvars table exists (no data is modified)
    - Returns the absolute store path
    """
    db_path = seqvar_store_path()
    db_path.parent.mkdir(parents=True, exist_ok=True)

    # Connect and apply schema in a single session
    try:
        with sqlite3.connect(db_path) as db:
            # Apply schema pragmas and table
            for stmt in filter(None, (s.strip() for s in SCHEMA_SQL.split(";"))):
                if stmt:
                    db.execute(stmt)
            # No data writes; just ensure structure exists
    except sqlite3.Error as e:
        raise RuntimeError(f"Failed to initialize seqvar store at {db_path}: {e}") from e

    return db_path


def _conn() -> sqlite3.Connection:
    p = seqvar_store_path()
    if not p.exists():
        # Create the store on demand
        init_seqvar_store()
    try:
        return sqlite3.connect(p)
    except sqlite3.Error as e:
        raise RuntimeError(f"cannot open seqvar store at {p}: {e}") from e


def get(key: str) -> str:
    with _conn() as db:
        row = db.execute("SELECT val FROM seqvars WHERE key=?", (key,)).fetchone()
        return "" if row is None or row[0] is None else row[0]


def set(key: str, val: str, src: str = None, as_default: bool = False) -> None:
    # If as_default is True and key already has a non-empty value, respect existing value
    if as_default and get(key):
        return

    now = int(time.time())

    # If src is not specified, try to get the caller function name
    if src is None:
        try:
            src = get_call_trace()
        except Exception:
            src = "??"

    with _conn() as db:
        db.execute(
            "INSERT INTO seqvars(key,val,src,ts) VALUES(?,?,?,?) "
            "ON CONFLICT(key) DO UPDATE SET val=excluded.val, src=excluded.src, ts=excluded.ts",
            (key, val, src, now)
        )


def dump() -> list[tuple]:
    """
    Return all rows and columns from the seqvar database.
    Returns a list of tuples, each containing (key, val, src, ts)
    """
    with _conn() as db:
        rows = db.execute("SELECT key, val, src, ts FROM seqvars ORDER BY key").fetchall()
        return rows


def get_dict(keys: str = None) -> dict[str, str]:
    """
    Return key-value pairs from the seqvar database as a dictionary.

    Args:
        keys: Optional filter pattern for keys. Uses SQLite LIKE operator with wildcards:
              - % matches any sequence of characters
              - _ matches any single character
              - Examples: "repos.*" -> "repos.%", "config.*.url" -> "config.%.url"

    Returns:
        Dictionary mapping keys to values

    Raises:
        RuntimeError: If there are database access issues
    """
    with _conn() as db:
        if keys is None:
            # No filter - get all key-value pairs
            rows = db.execute("SELECT key, val FROM seqvars ORDER BY key").fetchall()
        else:
            # Convert common wildcard patterns to SQLite LIKE patterns
            # Replace .* with .% for SQLite compatibility
            like_pattern = keys.replace(".*", ".%")
            rows = db.execute(
                "SELECT key, val FROM seqvars WHERE key LIKE ? ORDER BY key",
                (like_pattern,)
            ).fetchall()

        return {key: val for key, val in rows}


def set_dict(bindings: dict, src: str = None, as_default: bool = True) -> None:
    """
    Write multiple key-value pairs to the seqvar database.

    Args:
        bindings: Dictionary of key-value pairs to write
        src: Optional source identifier for the values
        as_default: If True, only set values for keys that are currently empty
          True is the default value because this is often used to fill in defaults.
    """
    for key, value in bindings.items():
        set(key, value, src, as_default)
